js/gatecoin.js

```diff
@@ -447,19 +450,41 @@ module.exports = class gatecoin extends Exchange {
                 throw new AuthenticationError (this.id + ' two-factor authentication requires a missing ValidationCode parameter');
         }
         let response = await this.privatePostTradeOrders (this.extend (order, params));
-        return {
+        let result = {
             'info': response,
             'id': response['clOrderId'],
         };
+        if ('responseStatus' in response) {
+            if ('message' in response['responseStatus']) {
+                if (response['responseStatus']['message'] === 'OK') {
+                    result['status'] = 'open';
+                }
+            }
+        }
+        return result;
     }
 
     async cancelOrder (id, symbol = undefined, params = {}) {
         await this.loadMarkets ();
-        return await this.privateDeleteTradeOrdersOrderID ({ 'OrderID': id });
+        let response = await this.privateDeleteTradeOrdersOrderID ({ 'OrderID': id });
+        if ('responseStatus' in response) {
+            if ('message' in response['responseStatus']) {
```

Comment:
> I.e. I am handling a successful case, not an error case.

Yep, I see that, however, what I'm telling you is exactly the opposite – you should handle the error case, not the successful case.

https://github.com/ccxt/ccxt/pull/2926/files#diff-0d6a31a134f8cae3c69f182bb5f62a25R471

↑ What will this method return if there's no `'responseStatus' in response` ? Or no `'message' in response['responseStatus']`. To me it looks like you really want to throw an error from handleErrors if responseStatus is missing or message !== 'OK'. In other words, we should revert the logic and handle what might go wrong in handleErrors, and not handle what might work well in all other methods. Does it make sense?

Reply:
To clarify: we want to catch 'Message' !== 'OK' in handleErrors and throw a specific error code, otherwise – throw nothing (yielding a good result).

Reply:
Thank you for the explanation! Indeed that makes sense.  I will submit another PR and close this one simultaneously. 

> What will this method return if there's no 'responseStatus' in response

Gotcha! I just love code reviews for that:)

Reply:
Well, nope, no luck here. 

See, gatecoin returns `responseStatus` and `responseStatus.message` only on certain [API](https://gatecoin.com/api/) calls so it isn't feasible to throw in `handleErrors` on absence of `responseStatus` in *all* of the responses. 

For instance in `cancelOrder` they only return `responseStatus.message` along with order id. So to figure out if the cancellation went successfully we have to look if `responseStatus.message == 'OK'`. What is not immediately apparent  from the source code is this: if the order being cancelled is not found then *this* case is handled by `handleErrors` via `this.exceptions` (I have added a few exceptions including the `OrderNotFound` which is  now thrown on `cancelOrder` in case it was already cancelled or did not exist in the first place). 

So it seems that we have to handle the successful case. 

BTW, every method in gatecoin.js is implemented just like this, by checking the successful case in `responseStatus.message` for method where this property is expected. 

What calls for improvement  though is this code which must be extracted in a separate method  in all occurrences: 

```javascript
        if ('responseStatus' in response)
            if ('message' in response['responseStatus'])
                if (response['responseStatus']['message'] === 'OK')
```

Do you approve? 

(Besides they don't document their error codes & `responseStatus`es, so everything in this PR including the updated statuses and updated exceptions I have deducted from real life trading.)

Reply:
> How does that sound to you?

Yes, this seems like a nice solution. Perfectly agree with that. 

> otherwise let it go through as being successful.

For some methods, yes. For `cancelOrder` I still have to parse the successful `response` and deduct if I can safely assume the order has indeed been canceled. And in that case I put the `response` in `info` and return the `.status`, `.id` and `.info` instead in the top-level. 

So let's do this: 

1. I put your snippet in `handleErrors`
2. I extract the success branch from all methods into a static parsing method
3. I review all methods and replace accordingly: either completely removing the successful branch from the method if all I need is to pass thru *or* checking the successful status and generate a response in appropriate cases. 

Sounds like a plan?

Reply:
Yep, thx for your understanding! 

I'm ok with a special case for `cancelOrder`, just need to make sure that we copy the rest of the handler properly. Use the already-existing implementations as examples. 

I'll be happy to answer your questions, if any.

Reply:
Thank you very much for your support and review!

> I'll be happy to answer your questions, if any.

What would be the best way to get in touch with you in that case?

Reply:
> What would be the best way to get in touch with you in that case?

You're already in touch with me in the best possible way here ) We also have a gitter:

[![Gitter](https://badges.gitter.im/ccxt-dev/ccxt.svg)](https://gitter.im/ccxt-dev/ccxt?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)

However, I'd prefer GitHub format anyway, to keep some separation of concerns )